src/Tasks.h

```diff
@@ -363,47 +431,93 @@ class PSSparseServerTask : public MLTask {
-          nworkers, worker_id, ps_ip, ps_port)
-  {}
-
-    /**
-     * Worker here is a value 0..nworkers - 1
-     */
-    void run(const Configuration& config, int worker);
+   MFNetflixTask(uint64_t model_size,
+                 uint64_t batch_size,
+                 uint64_t samples_per_batch,
+                 uint64_t features_per_sample,
+                 uint64_t nworkers,
+                 uint64_t worker_id,
+                 const std::string& ps_ip,
+                 uint64_t ps_port)
+       : MLTask(model_size,
+                batch_size,
+                samples_per_batch,
+                features_per_sample,
+                nworkers,
+                worker_id,
+                ps_ip,
+                ps_port) {
+     mf_model_get = std::make_unique<MFModelGet>(ps_ip, ps_port);
+   }
+   MFNetflixTask(uint64_t model_size,
+                 uint64_t batch_size,
+                 uint64_t samples_per_batch,
+                 uint64_t features_per_sample,
+                 uint64_t nworkers,
+                 uint64_t worker_id,
+                 std::vector<std::string> ps_ips,
+                 std::vector<uint64_t> ps_ports)
+       : MLTask(model_size,
+                batch_size,
+                samples_per_batch,
+                features_per_sample,
+                nworkers,
+                worker_id,
+                ps_ips,
+                ps_ports) {
+     mf_model_get =
+         std::make_unique<MFModelGet>(ps_ips, ps_ports, samples_per_batch);
+   }
+
+   /**
+    * Worker here is a value 0..nworkers - 1
+    */
+   void run(const Configuration& config, int worker);
 
   private:
-    class MFModelGet {
-      public:
-        MFModelGet(const std::string& ps_ip, int ps_port) :
-          ps_ip(ps_ip), ps_port(ps_port) {
-            psi = std::make_unique<PSSparseServerInterface>(ps_ip, ps_port);
-            psi->connect();
-        }
-
-        SparseMFModel get_new_model(const SparseDataset& ds,
-                                    uint64_t user_base_index,
-                                    uint64_t mb_size) {
-          return psi->get_sparse_mf_model(ds, user_base_index, mb_size);
-        }
-
-      private:
-        std::unique_ptr<PSSparseServerInterface> psi;
-        std::string ps_ip;
-        int ps_port;
+   class MFModelGet {
+    public:
+     MFModelGet(const std::string& ps_ip, int ps_port) {
+       psi = std::make_unique<PSSparseServerInterface>(ps_ip, ps_port);
+       while (true) {
+         try {
+           psi->connect();
+           break;
+         } catch (const std::exception& exc) {
+           std::cout << exc.what();
+         }
+       }
+     }
+
+     MFModelGet(const std::vector<std::string> ps_ips,
+                std::vector<uint64_t> ps_ports,
+                uint32_t minibatch_size) {
+       psi = std::make_unique<MultiplePSSparseServerInterface>(ps_ips, ps_ports,
+                                                               minibatch_size);
```

Comment:
For now, if you want to use MultiplePSSparseServerInterface on collaborative filtering, you need to construct it with minibatch_size. I don't know how to fix this. Ideally we would pass a Configuration object to MultiplePSSparseServerInterface, but that would require changing the PSSparseServerInterface in a lot of places.

Reply:
Not yet fixed.

Reply:
Fixed